### PR TITLE
cli, test, doc: bitcoin-cli -getinfo multiwallet balances follow-ups

### PR DESCRIPTION
**Pull request by jonatack**
These are follow-ups to #18594:

- release note for `-getinfo` multiwallet balances

- update the `-getinfo` help:
```
$ bitcoin-cli -h | grep -A3 getinfo
  -getinfo
       Get general information from the remote server, including the balances
       of each loaded wallet when in multiwallet mode. Note that
       -getinfo is the combined result of several RPCs (getnetworkinfo,
       getblockchaininfo, getwalletinfo, getbalances, and in multiwallet
       mode, listwallets), each with potentially different state.
```

- more robust bitcoin-cli -getinfo command parsing per review feedback in https://github.com/bitcoin/bitcoin/pull/18594#discussion_r428674779 and regression test coverage for that change

- add `assert_scale` to test_framework/util.py and improve the coverage in `interface_bitcoin_cli.py`

### CODE REVIEW

File: doc/release-notes-18594.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## CLI
+
+The `bitcoin-cli -getinfo` command now displays the wallet name and balance for
+each of the loaded wallets when more than one is loaded (e.g. in multiwallet
+mode) and a wallet is not specified with `-rpcwallet`.
```
Comment (MarcoFalke) [inline doc/release-notes-18594.md:5]:
```suggestion
mode) and a wallet is not specified with `-rpcwallet`. (#18594)
```

Comment (jonatack) [reply doc/release-notes-18594.md]: thanks, done

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -544,7 +544,9 @@ static int CommandLineRPC(int argc, char *argv[])
             args.erase(args.begin()); // Remove trailing method name from arguments vector
         }
         Optional<std::string> wallet_name{};
-        if (gArgs.IsArgSet("-rpcwallet")) wallet_name = gArgs.GetArg("-rpcwallet", "");
+        if (!gArgs.GetArgs("-rpcwallet").empty()) {
```
Comment (luke-jr) [inline src/bitcoin-cli.cpp:547]: What is this supposed to improve?

Comment (jonatack) [reply src/bitcoin-cli.cpp:547]: This is how `-rpcwallet` was parsed before (see diff in 7430775). I don't think it changes anything, but I mistakenly thought the same with respect to `-getinfo` before your feedback in that PR.

Comment (jonatack) [reply src/bitcoin-cli.cpp:547]: That said, I added `-rpcwallet` test coverage recently in #18724, so I think either way works.

Comment (jonatack) [reply src/bitcoin-cli.cpp]: Reverted to minimal change.

**Comment by hebasto** [comment]
Concept ACK.

File: test/functional/interface_bitcoin_cli.py [hunk 3]
```diff
@@ -13,6 +13,12 @@
 BLOCKS = 101
 BALANCE = (BLOCKS - 100) * 50
 
+def test_number_of_digits_after_decimal(amount):
```
Comment (jonatack) [inline test/functional/interface_bitcoin_cli.py:16]: For info, this test function is used in #19092 too.

**Review by hebasto — COMMENTED** [review]
Approach ACK 8019cf2dc7c4cf59ece71a6e7143b3f054a2243e.

@jonatack 
Could you consider to replace s/`ArgsManager::ALLOW_ANY`/`ArgsManager::ALLOW_BOOL`/ for `-getinfo` command-line argument?

There is a general direction to move from the `ArgsManager::ALLOW_ANY` flag to more specific ones:
- 7414d3820c833566b4f48c6c120a18bf53978c55 (`-rpcwhitelistdefault`)
- #16545

If you decide to implement this suggestion, the first commit and the last one could be combined.

**Comment by jonatack** [comment]
Thanks for reviewing, @hebasto! From what I can tell, it would be best to not change this (yet):

- `ALLOW_BOOL` appears to be unused in the codebase, apart from `-rpcwhitelistdefault`
- the change would have no effect on `-getinfo` command parsing behavior and tests
- in the event that we move to using more specific flags and that this one needs to be changed, many commands will need to be updated and ISTM it would be most coherent in terms of git history that they be changed together in the same commit or PR.

**Comment by hebasto** [comment]
> Thanks for reviewing, @hebasto! From what I can tell, it would be best to not change this (yet):
> 
>  - `ALLOW_BOOL` appears to be unused in the codebase, apart from `-rpcwhitelistdefault`
> 
>  - the change would have no effect on `-getinfo` command parsing behavior and tests
> 
>  - in the event that we move to using more specific flags and that this one needs to be changed, many commands will need to be updated and ISTM it would be most coherent in terms of git history that they be changed together in the same commit or PR.

FWIW, `ArgsManager::ALLOW_BOOL` is designed to use along with `ArgsManager::GetBoolArg()`. Additional error checking has not been implemented yet, unfortunately.

In any case, will continue my review :)

File: test/functional/test_framework/util.py [hunk 4]
```diff
@@ -32,6 +32,12 @@ def assert_approx(v, vexp, vspan=0.00001):
     if v > vexp + vspan:
         raise AssertionError("%s > [%s..%s]" % (str(v), str(vexp - vspan), str(vexp + vspan)))
 
+def assert_decimal_places(amount, decimal_places=8):
+    """Assert an amount has n digits (default 8 for amounts) after the decimal."""
+    remainder = str(amount).split('.')[-1]
+    if remainder != '0E-8':
```
Comment (hebasto) [inline test/functional/test_framework/util.py:38]: What cases are covered by this condition?

Comment (jonatack) [reply test/functional/test_framework/util.py:38]: The condition is for the case that there's no need to run the assert if the remainder string is the Python shorthand for the correct number of zero digits. But good catch: I should pass the param into it: s/`'0E-8'`/`'0E-{}'.format(decimal_places)`'

Comment (jonatack) [reply test/functional/test_framework/util.py]: Fixed and improved. Thanks @hebasto.

File: test/functional/test_framework/util.py [hunk 5]
```diff
@@ -180,6 +180,16 @@ def assert_array_result(object_array, to_match, expected, should_not_find=False)
     if num_matched > 0 and should_not_find:
         raise AssertionError("Objects were found %s" % (str(to_match)))
 
+def assert_scale(number, digits=8):
+    """Assert number has expected scale/fractional digits/number of digits after the decimal (default 8)."""
+    scale = str(number).split('.')[-1]
+    if scale == number:  # return early if number has no fractional part
+        return
```
Comment (hebasto) [inline test/functional/test_framework/util.py:187]: If a number has no fractional part, it should have `digits` zeros after the decimal, no?

**Review by hebasto — COMMENTED** [review]
Thanks for a separate commit that introduces the `assert_scale()` finction.

File: test/functional/test_framework/util.py (hunk 5 above)
Comment (jonatack) [reply test/functional/test_framework/util.py:187]: Good idea, done (thanks!)

File: test/functional/test_framework/util.py [hunk 6]
```diff
@@ -180,6 +180,18 @@ def assert_array_result(object_array, to_match, expected, should_not_find=False)
     if num_matched > 0 and should_not_find:
         raise AssertionError("Objects were found %s" % (str(to_match)))
 
+def assert_scale(number, expected_scale=8):
+    """Assert number has expected scale, e.g. fractional digits; number of
+    digits after the decimal. The default of 8 corresponds to a Bitcoin amount."""
+    number = str(number)
+    mantissa = number.split('.')[-1]
+    if mantissa[:3] == '0E-':
+        assert_equal(mantissa, '0E-{}'.format(expected_scale))  # exponent notation
```
Comment (hebasto) [inline test/functional/test_framework/util.py:189]: I'm not a python expert, but my python 3.6.9 prints numbers in scientific notation using "e" rather "E".
Btw, in what cases we could expect values with scientific notation in RPC responses?

Comment (jonatack) [reply test/functional/test_framework/util.py:189]: In every run of `interface_bitcoin_cli.py` the tests see zero values like `'paytxfee': Decimal('0E-8'),` in the -getinfo response. I see them as well if I add logging like `self.log.info(cli_get_info)` in the test.

Comment (jonatack) [reply test/functional/test_framework/util.py:189]: Example:
```
2020-06-06T16:11:48.077000Z TestFramework (INFO): Test -getinfo returns expected network and blockchain info
{
 'balance': Decimal('50.00000000'),
 'blocks': 101,
 'chain': 'regtest',
 'connections': 0,
 'difficulty': Decimal('4.656542373906925E-10'),
 'headers': 101,
 'keypoolsize': 1,
 'paytxfee': Decimal('0E-8'),
 'proxy': '',
 'relayfee': Decimal('0.00001000'),
 'timeoffset': 0,
 'unlocked_until': 0,
 'verificationprogress': 1,
 'version': 209900
}
```

Comment (jonatack) [reply test/functional/test_framework/util.py:189]: Added `upper()` just in case.

**Comment by jonatack** [comment]
Happy to see this is in Bitcoin Knots. Rebased.

**Comment by jonatack** [comment]
Rebased!

**Comment by jonatack** [comment]
Closing, will propose the changes one by one.

**Pull request closed without merging**